src/coreclr/src/vm/tailcallhelp.cpp

```diff
@@ -0,0 +1,871 @@
+    InterlockedCompareExchangeT(&s_tailCallDispatcherMD, pDispatchTailCallsMD, NULL);
+    return s_tailCallDispatcherMD;
+}
+
+void TailCallHelp::CreateTailCallHelperStubs(
+    MethodDesc* pCallerMD, MethodDesc* pCalleeMD,
+    MetaSig& callSiteSig, bool virt, bool thisArgByRef,
+    MethodDesc** storeArgsStub, bool* storeArgsNeedsTarget,
+    MethodDesc** callTargetStub)
+{
+    STANDARD_VM_CONTRACT;
+
+#ifdef _DEBUG
+    SigFormat incSig(callSiteSig, NULL);
+    LOG((LF_STUBS, LL_INFO1000, "TAILCALLHELP: Incoming sig %s\n", incSig.GetCString()));
+#endif
+
+    *storeArgsNeedsTarget = pCalleeMD == NULL || pCalleeMD->IsSharedByGenericInstantiations();
+
+    // The tailcall helper stubs are always allocated together with the caller.
+    // If we ever wish to share these stubs they should be allocated with the
+    // callee in most cases.
+    LoaderAllocator* pLoaderAllocator = pCallerMD->GetLoaderAllocator();
+
+    TypeHandle retTyHnd = NormalizeSigType(callSiteSig.GetRetTypeHandleThrowing());
+    TailCallInfo info(pCallerMD, pCalleeMD, pLoaderAllocator, &callSiteSig, virt, retTyHnd);
+
+    LayOutArgBuffer(callSiteSig, pCalleeMD, *storeArgsNeedsTarget, thisArgByRef, &info.ArgBufLayout);
+    info.HasGCDescriptor = GenerateGCDescriptor(pCalleeMD, info.ArgBufLayout, &info.GCRefMapBuilder);
+
+    *storeArgsStub = CreateStoreArgsStub(info);
+    *callTargetStub = CreateCallTargetStub(info);
+}
+
+void TailCallHelp::LayOutArgBuffer(
+    MetaSig& callSiteSig, MethodDesc* calleeMD,
+    bool storeTarget, bool thisArgByRef, ArgBufferLayout* layout)
+{
+    unsigned int offs = 0;
+
+    auto addValue = [&](TypeHandle th)
+    {
+        unsigned int alignment = CEEInfo::getClassAlignmentRequirementStatic(th);
+        offs = AlignUp(offs, alignment);
+        layout->Values.Append(ArgBufferValue(th, offs));
+        offs += th.GetSize();
+    };
+
+    // User args
+    if (callSiteSig.HasThis() && !callSiteSig.HasExplicitThis())
+    {
+        TypeHandle thisHnd;
+
+        bool thisParamByRef = (calleeMD != NULL) ? calleeMD->GetMethodTable()->IsValueType() : thisArgByRef;
+        if (thisParamByRef)
+        {
+            thisHnd = TypeHandle(MscorlibBinder::GetElementType(ELEMENT_TYPE_U1))
+                      .MakeByRef();
+        }
+        else
+        {
+            thisHnd = TypeHandle(g_pObjectClass);
+        }
+
+        addValue(thisHnd);
+    }
+
+    callSiteSig.Reset();
+    CorElementType ty;
+    while ((ty = callSiteSig.NextArg()) != ELEMENT_TYPE_END)
+    {
+        TypeHandle tyHnd = callSiteSig.GetLastTypeHandleThrowing();
+        tyHnd = NormalizeSigType(tyHnd);
+        addValue(tyHnd);
+    }
+
+    if (storeTarget)
+    {
+        offs = AlignUp(offs, TARGET_POINTER_SIZE);
+        layout->TargetAddressOffset = offs;
+        layout->HasTargetAddress = true;
+        offs += TARGET_POINTER_SIZE;
+    }
+
+    layout->Size = offs;
+}
+
+TypeHandle TailCallHelp::NormalizeSigType(TypeHandle tyHnd)
```

Comment:
Why do we need to normalize the type handles in the signatures? Comment would be useful.

Reply:
@jakobbotsch Is this done to make it easier to share stubs in the future?

Reply:
The main reason is so that `StoreTailCallArgs` stub does not need to be generic when the target function is generic, I believe. I will add a comment.